File: enode/src/main/java/org/enodeframework/queue/applicationmessage/DefaultApplicationMessageHandler.java
package org.enodeframework.queue.applicationmessage;

import org.enodeframework.common.serializing.SerializeService;
import org.enodeframework.infrastructure.TypeNameProvider;
import org.enodeframework.messaging.ApplicationMessage;
import org.enodeframework.messaging.MessageDispatcher;
import org.enodeframework.queue.MessageContext;
import org.enodeframework.queue.MessageHandler;
import org.enodeframework.queue.QueueMessage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class DefaultApplicationMessageHandler implements MessageHandler {
    private static final Logger logger = LoggerFactory.getLogger(DefaultApplicationMessageHandler.class);
    private final TypeNameProvider typeNameProvider;
    private final MessageDispatcher messageDispatcher;
    private final SerializeService serializeService;

    public DefaultApplicationMessageHandler(TypeNameProvider typeNameProvider, MessageDispatcher messageDispatcher, SerializeService serializeService) {
        this.typeNameProvider = typeNameProvider;
        this.messageDispatcher = messageDispatcher;
        this.serializeService = serializeService;
    }

    @Override
    public void handle(QueueMessage queueMessage, MessageContext context) {
        logger.info("Received application message: {}", serializeService.serialize(queueMessage));
        String msg = queueMessage.getBody();
        GenericApplicationMessage appDataMessage = serializeService.deserialize(msg, GenericApplicationMessage.class);
        Class<?> applicationMessageType = typeNameProvider.getType(appDataMessage.getApplicationMessageType());
        ApplicationMessage message = (ApplicationMessage) serializeService.deserialize(appDataMessage.getApplicationMessageData(), applicationMessageType);
        messageDispatcher.dispatchMessageAsync(message).whenComplete((x, y) -> {
            context.onMessageHandled(queueMessage);
        });
    }
}

File: enode/src/main/java/org/enodeframework/queue/applicationmessage/DefaultApplicationMessageHandler.kt
package org.enodeframework.queue.applicationmessage

import org.enodeframework.common.serializing.SerializeService
import org.enodeframework.infrastructure.TypeNameProvider
import org.enodeframework.messaging.ApplicationMessage
import org.enodeframework.messaging.MessageDispatcher
import org.enodeframework.queue.MessageContext
import org.enodeframework.queue.MessageHandler
import org.enodeframework.queue.QueueMessage
import org.slf4j.LoggerFactory

class DefaultApplicationMessageHandler(
    private val typeNameProvider: TypeNameProvider,
    private val messageDispatcher: MessageDispatcher,
    private val serializeService: SerializeService
) : MessageHandler {
    override fun handle(queueMessage: QueueMessage, context: MessageContext) {
        logger.info("Received application message: {}", queueMessage)
        val appDataMessage = serializeService.deserialize(queueMessage.body, GenericApplicationMessage::class.java)
        val applicationMessageType = typeNameProvider.getType(appDataMessage.applicationMessageType)
        val message = serializeService.deserialize(
            appDataMessage.applicationMessageData,
            applicationMessageType
        ) as ApplicationMessage
        messageDispatcher.dispatchMessageAsync(message)
            .whenComplete { x: Boolean?, y: Throwable? -> context.onMessageHandled(queueMessage) }
    }

    companion object {
        private val logger = LoggerFactory.getLogger(DefaultApplicationMessageHandler::class.java)
    }
}
